### PR TITLE
Overmap Saving

### PR DESCRIPTION
**Pull request by NataKilar**
Overmap is now generated using a random map that designates areas as hazardous or non-hazardous, with the former having random overmap events being generated on the tiles. The intent of this is to create natural "space-lanes", with travel outside of the winding path being risky, but quicker. The random map uses a seed, so it will remain the same throughout save/load. The specific types of hazards, however, will change with each load. The away map system has not been integrated into this yet.

Overmap sectors and ships, both landable and non-landable, will now save. The determination of whether ships and sectors will save is dependent on their z-level(s) being saved, as the system moves the sectors to their mapped z-levels prior to save.

A small note is that currently the visitor shuttle landmarks are not saved. A replacement system is something I'm currently working on.

On the (further) backend, areas that are saved but not in a saved z-level (currently, landable ships) will properly be reconstructed. Datums with vars that point to a object that requires a wrapper will now point to a single wrapper for each instance. This could probably be done cleaner with a separate table in the database, but that may be outside of my ability. This is particularly important for saving areas, as created areas are all of the type /area/, so locate cannot be used on them, and therefore differentiating between when an area should be created fresh and when tiles should just be added to the area is impossible in the current implementation. This lead to a bug in which every reference to an area would create a new instance of type /area/ and re-assign the tiles to the new instance for every reference.

### CODE REVIEW

File: mods/persistence/controllers/subsystems/persistence.dm [hunk 1]
```diff
@@ -161,13 +158,37 @@ SUBSYSTEM_DEF(persistence)
...
+				// Batch them up to save time.
+				if(index % 128 == 0)
+					serializer.Commit()
+					index = 1
+				else
+					index++
+
+				// Prevent the whole game from locking up.
+				CHECK_TICK
+			serializer.Commit() // cleanup leftovers.
```
Comment (NataKilar) [inline mods/persistence/controllers/subsystems/persistence.dm:191]: This particular section needs to be wholly reviewed and likely replaced, as it's almost a 1:1 copy paste from the serialization code above.

File: mods/persistence/controllers/subsystems/mapping.dm [hunk 2]
```diff
@@ -17,20 +17,46 @@
...
 	SSpersistence.LoadWorld()
+	
+	// Initialize the overmap for Persistence.
+	if(GLOB.using_map.use_overmap)
+		if(!GLOB.using_map.overmap_z)
+			build_overmap() // If a overmap hasn't been loaded, create a new one.
+
+	// Generate the areas of the overmap that are hazardous.
+	// Map is placed at the lower left corner of the overmap, not including the edges (2, 2).
+	new /datum/random_map/automata/overmap(null,2,2,GLOB.using_map.overmap_z,GLOB.using_map.overmap_size-2,GLOB.using_map.overmap_size-2,0,1,1)
```
Comment (biscuitWizard) [inline mods/persistence/controllers/subsystems/mapping.dm:42]: reeee magic numbers

why -2?

File: mods/persistence/controllers/subsystems/persistence.dm [hunk 3]
```diff
@@ -27,6 +27,10 @@ SUBSYSTEM_DEF(persistence)
 	// Collect the z-levels we're saving and get the turfs!
 	to_world_log("Saving [LAZYLEN(SSpersistence.saved_levels)] z-levels. World size max ([world.maxx],[world.maxy])")
 	var/start = world.timeofday
+	
+	// Launch events
+
+	GLOB.world_saving_start_event.raise_event(src)
```
Comment (biscuitWizard) [inline mods/persistence/controllers/subsystems/persistence.dm:33]: Why was this moved to before the wipe? Out of curiosity.

File: mods/persistence/modules/overmap/landable.dm [hunk 4]
```diff
@@ -1,6 +1,49 @@
...
+			map_z = landmark.z
+			return
+	// Otherwise, the ship has likely landed elsewhere, and needs to reconstruct its space z-level.
+	qdel(landmark)
+	. = ..()
+
+// The landable ship contains a reference to its landmark, so only save if the ship is in its z-level.
+/obj/effect/shuttle_landmark/ship/should_save()
+	var/datum/shuttle/S = SSshuttle.shuttles[shuttle_name]
+	if(S && S.current_location && S.current_location == src)
```
Comment (biscuitWizard) [inline mods/persistence/modules/overmap/landable.dm:27]: Simplify to:
`return S && S.current_location && S.current_location == src`

File: mods/persistence/modules/world_save/serializers/json_serializer.dm [hunk 5]
```diff
@@ -31,8 +31,10 @@
 				results[V] = "FLAT_OBJ#[SerializeDatum(VV)]"
 			else
 				results[V] = "OBJ#[sql.SerializeDatum(VV)]"
+	// If the object is a wrapper, keep track of the reference to the wrapped object rather than the wrapper itself.
+	var/thing_id = object_parent ? "\ref[object_parent]" : "\ref[object]"
```
Comment (biscuitWizard) [inline mods/persistence/modules/world_save/serializers/json_serializer.dm]: What?

This seems anomalous, why is this happening here and what is it supposed to do? Later thing_id is used for the reverse mapper, but \ref[] returns a memory address, and all entries in reverse_map are integer-strings. This breaks convention.

File: mods/persistence/modules/world_save/serializers/json_serializer.dm [hunk 6]
```diff
@@ -77,7 +79,12 @@
 /serializer/json/DeserializeDatum(var/datum/persistence/load_cache/thing/object)
 	throw EXCEPTION("Do not use DeserializeDatum for the JSON Serializer. Use QueryAndDeserializeDatum.")
 
-/serializer/proc/JsonDeserializeDatum(var/datum/thing, var/list/thing_vars)
+/serializer/proc/JsonDeserializeDatum(var/thing_type, var/list/thing_vars, var/thing_id)
+	var/datum/thing = reverse_map["[thing_id]"]
```
Comment (biscuitWizard) [inline mods/persistence/modules/world_save/serializers/json_serializer.dm]: Why is a memory address being assigned in reverse_map? The key for reverse_map is always the index of the table row for that object.

File: mods/persistence/modules/world_save/wrappers/_wrapper.dm [hunk 7]
```diff
@@ -1,9 +1,11 @@
 /datum/wrapper
 	var/key
 	var/wrapper_for
+	var/deserialized_thing // the thing that this wrapper has already deserialized, if a new instance is not desired for every reference to the wrapper.
```
Comment (biscuitWizard) [inline mods/persistence/modules/world_save/wrappers/_wrapper.dm]: This is scoped incorrectly and should probably just only be on the area wrapper, since this isn't generically assigned to at the /datum/wrapper level.

**Review by biscuitWizard — CHANGES_REQUESTED** [review]
Some architectural questions about the modifications to the serialization system. Extra guarded, as any changes to said system need to be vetted strictly given the complexity of the system.

File: mods/persistence/controllers/subsystems/mapping.dm (hunk 2 above)
Comment (NataKilar) [reply mods/persistence/controllers/subsystems/mapping.dm:42]: The edges of the overmap are composed of a different tile that is used for looping. Because of this, the edges can't have the overmap automata overlaid on top of it, so the -2, -2 while applying the map at 2, 2 just ignores the edges.

File: mods/persistence/controllers/subsystems/persistence.dm (hunk 3 above)
Comment (NataKilar) [reply mods/persistence/controllers/subsystems/persistence.dm:33]: Just for consistency's sake, as the end event occurs outside of the try/catch. Neutral on it though, and it can be moved back if it makes more sense to you.

File: mods/persistence/modules/world_save/serializers/json_serializer.dm (hunk 5 above)
Comment (NataKilar) [reply mods/persistence/modules/world_save/serializers/json_serializer.dm]: The use of \ref[] here is arbitrary; it's just to get a unique identifier for the datum to ensure that only one wrapper per object is created on de-serialization (which itself has issues, but in general this approach was a bit makeshift). The issue that arises is that because flattened objects are not otherwise saved, there isn't any index to use for the key.

Needless to say though, this is pretty lackluster even if it 'works'. A possible alternative would be to add another table which contains the wrappers, and change the JSON serializer to use thing_map to make sure that each reference to an object only points to the single wrapper. I have next to 0 experience with SQL or databases in general however, and I'm not sure of the potential performance impact this would have as opposed to other solutions.

**Comment by NataKilar** [comment]
Went ahead and tore out the changes to the serialization system and replaced them according to our discussion on discord. Initial testing shows it seems to work just as well however. Need to get some additional live testing done on this probably, but should be good to go.
